apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabaseBase.cs

```diff
@@ -0,0 +1,224 @@
+                : EmulatorDetection.EmulatorOrProduction
+        };
+        if (SpannerPort != null)
+        {
+            builder.Port = int.Parse(SpannerPort);
+        }
+        NoDbConnectionString = builder.ConnectionString;
+        SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
+        var databaseBuilder = builder.WithDatabase(SpannerDatabase);
+        ConnectionString = databaseBuilder.ConnectionString;
+        DatabaseName = databaseBuilder.DatabaseName;
+
+        if (dialect != DatabaseDialect.Postgresql)
+        {
+            // We create database instance only on the emulator.
+            // In production, we expect the instance to already exist.
+            MaybeCreateInstanceOnEmulator(projectId);
+        }
+
+        if (Fresh)
+        {
+            CreateDatabase(dialect);
+            Logger.DefaultLogger.Debug($"Created database {SpannerDatabase} with {dialect} dialect.");
+        }
+        else
+        {
+            Logger.DefaultLogger.Debug($"Using existing database {SpannerDatabase}");
+        }
+    }
+
+    protected static string GetEnvironmentVariableOrDefault(string name, string defaultValue)
+    {
+        string value = Environment.GetEnvironmentVariable(name);
+        return string.IsNullOrEmpty(value) ? defaultValue : value;
+    }
+
+    protected virtual void CreateDatabase(DatabaseDialect dialect)
+    {
+        // PostgreSQL database can only be created by Admin APIs, which is not supported by Emulator for GSQL database creation.
+        // So, we need two different implementation for GSQL and PostgreSQL.
+        if (dialect == DatabaseDialect.Postgresql)
+        {
+            DatabaseAdminClient databaseAdminClient = DatabaseAdminClient.Create();
+            CreateDatabaseRequest createDatabaseRequest = new CreateDatabaseRequest
+            {
+                CreateStatement = $"CREATE DATABASE {SpannerDatabase}",
+                DatabaseDialect = dialect,
+                ParentAsInstanceName = InstanceName.FromProjectInstance(ProjectId, SpannerInstance)
+            };
+
+            try
+            {
+                var operation = databaseAdminClient.CreateDatabase(createDatabaseRequest);
+                var completedResponse = operation.PollUntilCompleted();
+                if (completedResponse.IsFaulted)
+                {
+                    Logger.DefaultLogger.Debug($"Error while creating database with dialect {dialect}: {completedResponse.Exception}");
+                    throw completedResponse.Exception;
+                }
+
+                Logger.DefaultLogger.Debug($"The database {SpannerDatabase} with dialect {dialect} created successfully.");
+            }
+            catch (RpcException e) when (e.StatusCode == StatusCode.AlreadyExists)
+            {
+                // Ignore.
+            }
+        }
+        else
+        {
+            using var connection = new SpannerConnection(NoDbConnectionString);
```

Comment:
And everything on this else should be the implementation of this method on the GoogleDialect specific derived class.